src/compiler/checker.ts

```diff
@@ -38599,13 +38607,52 @@ namespace ts {
                     error(node, message, symbolToString(symbol));
                 }
 
-                // Don't allow to re-export something with no value side when `--isolatedModules` is set.
+                // Type assertion to defeat the type predicate aliasing of `node` here because
+                // `isTypeOnlyImportOrExportDeclaration` is a one-sided type guard so `!isDeclaredTypeOnly`
+                // does not actually imply any narrowing effect on `node`.
+                const isDeclaredTypeOnly = isTypeOnlyImportOrExportDeclaration(node as never);
                 if (compilerOptions.isolatedModules
-                    && node.kind === SyntaxKind.ExportSpecifier
-                    && !node.parent.parent.isTypeOnly
-                    && !(target.flags & SymbolFlags.Value)
-                    && !(node.flags & NodeFlags.Ambient)) {
-                    error(node, Diagnostics.Re_exporting_a_type_when_the_isolatedModules_flag_is_provided_requires_using_export_type);
+                    && !isDeclaredTypeOnly
+                    && !(node.flags & NodeFlags.Ambient)
+                    && (!(target.flags & SymbolFlags.Value) || getTypeOnlyAliasDeclaration(symbol))) {
+                    const isType = !(target.flags & SymbolFlags.Value);
+                    const typeOnlyAlias = getTypeOnlyAliasDeclaration(symbol);
+

```

Comment:
Would it make sense to use a nested `if` here to avoid repetition?

```suggestion
                    && !(node.flags & NodeFlags.Ambient)) {
                    const isType = !(target.flags & SymbolFlags.Value);
                    const typeOnlyAlias = getTypeOnlyAliasDeclaration(symbol);
                    if (isType || typeOnlyAlias) {
```